abyss: import from homebrew/science

jonchang opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Science PR: https://github.com/Homebrew/homebrew-science/pull/6332

c.f. https://github.com/Homebrew/homebrew-science/issues/6331

jonchang commented:
```
Error: 1 problem in 1 formula
abyss:
  * Formula name conflicts with caskroom/cask/abyss, homebrew/science/abyss
```

Not sure what to do about this one. CI passes otherwise.

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+    depends_on "multimarkdown" => :build
+  end
+
+  option :cxx11
```
shouldn't need this.

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  option :cxx11
+  option "with-maxk=", "Set the maximum k-mer length to N [default is 128]"
```
can we pick a sensible default and not make this an option

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+
+  option :cxx11
+  option "with-maxk=", "Set the maximum k-mer length to N [default is 128]"
+  option "without-test", "Skip build-time tests (not recommended)"
```
no running `make check` or `make test` in homebrew/core unless it's security software

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+
+  needs :openmp if build.with? "openmp"
+
+  # Only header files are used from these packages, so :build is appropriate
```
this can be assumed based on their `=> :build` status so comment isn't needed

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+  # Only header files are used from these packages, so :build is appropriate
+  depends_on "boost" => :build
+  depends_on "google-sparsehash" => :build
+  depends_on :mpi => [:cc, :recommended]
```
don't think we need the opt out.

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+  depends_on :mpi => [:cc, :recommended]
+
+  # strip breaks the ability to read compressed files.
+  skip_clean "bin"
```
don't think we need this

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+  url "https://github.com/bcgsc/abyss/releases/download/2.0.2/abyss-2.0.2.tar.gz"
+  sha256 "d87b76edeac3a6fb48f24a1d63f243d8278a324c9a5eb29027b640f7089422df"
+
+  head do
```
don't think we need the head spec. no one has used it in the last 90 days

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  test do
+    system "#{bin}/ABYSS", "--version"
```
needs a real test

jonchang commented:
Addressed all comments except the test. Is it OK to have a resource for just test data? The README.md links to one but it's not included in the distribution. Pinging @sjackman in case he has a smaller dataset.

ilovezfs commented:
>Is it OK to have a resource for just test data?

Yup.
```
Josephs-MacBook-Pro:Formula joe$ ag testpath.install|grep resource|wc -l
      13
Josephs-MacBook-Pro:Formula joe$ ag testpath|grep resource|grep stage|wc -l
      20
```

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,32 @@
+  def install
+    system "./autogen.sh" if build.head?
+
+    args = [
```
These arguments should be inlined since nothing is optionally appended to the array.

ilovezfs replied:
But, do we need to use `--disable-openmp` when building without it?

jonchang replied:
It checks for openmp and disables it since clang doesn't support it. I've added it in regardless.

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,32 @@
+  depends_on :mpi => :cc
+
+  def install
+    system "./autogen.sh" if build.head?
```
head is dead

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,32 @@
+    args = [
+      "--enable-maxk=128",
+      "--prefix=#{prefix}",
+      "--disable-dependency-tracking",
```
also --disable-silent-rules

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,32 @@
+
+  option "with-openmp", "Enable OpenMP multithreading"
+
+  needs :openmp if build.with? "openmp"
```
almost no one is opting out (3 out of 102 installs in the last 90 days). Can we just kill the option?

ilovezfs replied:
sorry I mean no one is using the option at all. I always get mpi and openmp confused.
```
Josephs-MacBook-Pro:~ joe$ brew formula-analytics --days-ago=90 abyss
install events in the last 90 days for abyss
=====================================================================================================
1 | homebrew/science/abyss                                                            |  98 |  96.08%
2 | homebrew/science/abyss --without-mpi                                              |   2 |   1.96%
3 | homebrew/science/abyss --c++11                                                    |   1 |   0.98%
4 | homebrew/science/abyss --c++11 --without-mpi                                      |   1 |   0.98%
=====================================================================================================
Total                                                                                 | 102 |    100%
=====================================================================================================
```
I don't think we need the option if no one is using it.

jonchang commented:
* took out all options
* added a real test

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented:
@vitorgalvao any thoughts on how we should handle the name conflict with https://github.com/caskroom/homebrew-cask/blob/master/Casks/abyss.rb ?

sjackman commented:
The cask app is named `Abyss Web Server.app`. I'd suggest renaming the cask to `abyss-web-server`.

sjackman commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,30 @@
+  def install
+    system "./configure", "--enable-maxk=128", "--prefix=#{prefix}",
+           "--disable-dependency-tracking", "--disable-silent-rules",
+           "--disable-openmp"
```
Please add `needs :openmp` and `--enable-openmp`. The majority of the ABySS pipeline is parallelized using OpenMP. Only the very first stage is parallelized using MPI.

ilovezfs replied:
```
  option "with-openmp", "Enable OpenMP multithreading"
```
was used by no one in the last 90 days so why would we make it the default?

sjackman replied:
Because otherwise the tool would not be parallelized, which would make assembly of large genomes terribly slow. Users use the default because it's the default, and they do not know better.

ilovezfs replied:
And you're OK with no opt-out?

jonchang replied:
I’ll defer to Shaun on this once since I’ve only used abyss once or twice. I’m ok with adding `needs :openmp` as a default.

ilovezfs replied:
:+1:

sjackman replied:
Yes, `needs :openmp` as the default with no opt-out is fine.

sjackman commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,30 @@
+  desc "Genome sequence assembler for short reads"
+  homepage "http://www.bcgsc.ca/platform/bioinfo/software/abyss"
+  url "https://github.com/bcgsc/abyss/releases/download/2.0.2/abyss-2.0.2.tar.gz"
+  sha256 "d87b76edeac3a6fb48f24a1d63f243d8278a324c9a5eb29027b640f7089422df"
```
Are `head` being intentionally removed?

ilovezfs replied:
Yeah, 0 uses of it in the last 90 days.

sjackman replied:
I use `--HEAD` as a mode for a user to test a fix of a bug that's affecting them before tagging a release. No bugs in the last ninety days would mean no installs of `--HEAD` in the last ninety days.

ilovezfs replied:
It's also 0 at 30, 60, 90, 120, 150, 180, 210, 240, ...

sjackman replied:
My preference is to include `head`. Final call is yours of course.

ilovezfs replied:
Sure let's include it. But just this once since you're upstream :p

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,30 @@
+           "--disable-dependency-tracking", "--disable-silent-rules",
+           "--disable-openmp"
+
+    system "make"
```
Is the separate `make` step actually needed?

sjackman replied:
No.

sjackman commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,30 @@
+    system "make"
+    system "make", "install"
+  end
+
```
Please add…
```ruby
  def caveats; <<-EOS.undent
    Please cite...
      Jackman et al. (2017) http://doi.org/10.1101/gr.214346.116
      Simpson et al. (2009) http://doi.org/10.1101/gr.089532.108
    EOS
  end
```

sjackman replied:
Okay.

jonchang replied:
Shaun, if my fellowship proposal is accepted I'm planning on writing a `brew cite` external command 🙂

sjackman replied:
That'd be awesome.

sjackman commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,30 @@
+  test do
+    testpath.install resource("testdata")
+    system "#{bin}/abyss-pe", "k=25", "name=test", "in=reads1.fastq reads2.fastq"
+    system "#{bin}/abyss-fac", "test-unitigs.fa"
```
Thanks for adding the test!

jonchang commented:
* `needs :openmp` is now default sans opt out
* `make install` is just one step

Built and tested locally w/ xcode 9 sierra. Hoping this is good to :ship: once CI is 🍏

ilovezfs commented:
>The cask app is named Abyss Web Server.app. I'd suggest renaming the cask to abyss-web-server.

@jonchang can you open a PR to https://github.com/caskroom/homebrew-cask for the rename please?

jonchang commented:
* head spec is back from the dead

ilovezfs commented on Formula/abyss.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def install
+    system "./autogen.sh" if build.head?
+    system "./configure", "--enable-maxk=128", "--prefix=#{prefix}",
```
configure args should be stacked as in the `brew create` template

ilovezfs replied:
```
        system "./configure", "--disable-dependency-tracking",
                              "--disable-silent-rules",
                              "--prefix=#{prefix}",
                              "--enable-maxk=128"
```

jonchang commented:
* configure args appropriately stacked

ilovezfs commented:
Thanks @jonchang!

sjackman commented:
Thanks, Jon!

vitorgalvao commented:
I’ve been away for a few days, hence why I haven’t replied sooner. Seems like my input is no longer needed, but if it is please let me know.

sjackman commented:
No worries, Vítor. `abyss` was renamed to `abyss-web-server` in PR https://github.com/caskroom/homebrew-cask/pull/38925.
